chains/sequential.go

```diff
@@ -0,0 +1,205 @@
+package chains
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/tmc/langchaingo/memory"
+	"github.com/tmc/langchaingo/schema"
+)
+
+// SequentialChain is a chain that runs multiple chains in sequence,
+// where the output of one chain is the input of the next.
+type SequentialChain struct {
+	chains     []Chain
+	inputKeys  []string
+	outputKeys []string
+	memory     schema.Memory
+}
+
+func NewSequentialChain(chains []Chain, inputKeys []string, outputKeys []string) (*SequentialChain, error) {
+	if err := validateSeqChain(chains, inputKeys, outputKeys); err != nil {
+		return nil, err
+	}
+
+	return &SequentialChain{
+		chains:     chains,
+		inputKeys:  inputKeys,
+		outputKeys: outputKeys,
+		memory:     memory.NewSimple(),
+	}, nil
+}
+
+func validateSeqChain(chain []Chain, inputKeys []string, outputKeys []string) error {
+	knownKeys := toSet(inputKeys)
+
+	for i, c := range chain {
+		// Check that chain has input keys that are in knownKeys
+		missingKeys := difference(c.GetInputKeys(), knownKeys)
+		if len(missingKeys) > 0 {
+			return fmt.Errorf(
+				"%w: chain at index %d is missing required input keys: %v",
+				ErrChainInitialization, i, missingKeys,
+			)
+		}
+
+		// Check that chain does not have output keys that are already in knownKeys
+		overlappingKeys := intersection(c.GetOutputKeys(), knownKeys)
+		if len(overlappingKeys) > 0 {
+			return fmt.Errorf(
+				"%w: chain at index %d has output keys that already exist: %v",
+				ErrChainInitialization, i, overlappingKeys,
+			)
+		}
+
+		// Add the chain's output keys to knownKeys
+		for _, key := range c.GetOutputKeys() {
+			knownKeys[key] = struct{}{}
+		}
+	}
+
+	// Check that outputKeys are in knownKeys
+	for _, key := range outputKeys {
+		if _, ok := knownKeys[key]; !ok {
+			return fmt.Errorf("%w: output key %s is not in the known keys", ErrChainInitialization, key)
+		}
+	}
+
+	return nil
+}
+
+// toSet converts a list to a set.
+func toSet(list []string) map[string]struct{} {
```

Comment:
What do u think about putting such utility functions inside a separate package named util or helper to be usable in other places? They seem to be general.

Reply:
My preference is to leave these functions close to where they are used until another use-case needs them. I do see your point though - maybe the maintainers can chime in here.

Reply:
I thought about this again and moved the functions to a `util` package. Given the rate of change of this project, it's probably worth bubbling up a `util` package that others can see and can take advantage of if needed.

Thanks 👍🏾